### PR TITLE
Fixes #586 TLS redirection errors.

### LINKED ISSUE #586: HTTPS issue with www.quick-lint-js.com
https://www.quick-lint-js.com/

Error code: SSL_ERROR_BAD_CERT_DOMAIN

### PR DESCRIPTION
**Pull request by 70m4c**
This submission is made as a draft because I have no reasonable way of testing this configruation and am submitting it blind.

This is an attempt to fix #586. There is currently no virtual host block that defines `ServerName www.quick-lint-js.com`. As a result, Apache is relying on its default behavior of using the first-defined virtual server blocks, which are currently defined in `001-quicklintjs.conf`.

The present behavior of loading http://www.quick-lint-js.com is appropriate for web browsers that do not automatically upgrade to a secure connection. Using the configuration for port 80 in `001-quicklintjs.conf`, the browser will correctly redirect to https://quick-lint-js.com. However, if a browser attempts to upgrade the connection, the user will be redirected to https://www.quick-lint-js.com.

The behavior of https://www.quick-lint-js.com is incorrect because it loads the TLS certificates located in
- `/etc/letsencrypt/live/quicklintjs.com/fullchain.pem` and
- `/etc/letsencrypt/live/quicklintjs.com/privkey.pem`

rather than

- `/etc/letsencrypt/live/quick-lint-js.com/fullchain.pem` and 
- `/etc/letsencrypt/live/quick-lint-js.com/privkey.pem`.

This behavior is due (as before) to there being no defined virtual host block for `www.quick-lint-js.com` on port 443 and Apache using the first virtual host block for port 443 defined in `001-quicklintjs.conf`. Because of the early TLS error, the user is never able to reach the later HTTP redirection stage.

In order to help prevent future situations that create confusion due to the use of the default virtual host block, I also crated an experimental (and untested) set of default blocks that attempt to mimic the behavior of Nginx's 444 Connection Closed Without Response error. The idea behind this approach is to refuse serving any site that is not explicitly defined in a virtual host block. That avoids unclear situations like #586 and may prevent the web server from participating in any attempts by a third part to point a domain to the server and to have it respond using the default block.

I also renamed some files for consistency.

### CODE REVIEW

**Comment by 70m4c** [comment]
I set up a couple domains I own but don't use to test this config. I'll let the DNS propagate over night and test tomorrow.

File: infrastructure/quick-lint-js-web/apache2/sites-available/000-default.conf [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# This is the default VirtualHost.
+
+# These blocks attempt to create an equivalent response to Nginx's non-standard
+# error 444 Connection Closed Without Response per the suggestion here:
+# https://serverfault.com/questions/822484/nginx-444-error-eqivalent-in-apache#comment1047893_822484
+# These blocks should only be matched when no other block is matched.
+<VirtualHost *:80>
+	ServerName default
+	ErrorDocument 444 "Connection Closed Without Response"
```
Comment (strager) [inline infrastructure/quick-lint-js-web/apache2/sites-available/000-default.conf:9]: Must fix:
```
AH00526: Syntax error on line 9 of /etc/apache2/sites-enabled/000-default.conf:
Unsupported HTTP response code 444
```

File: infrastructure/quick-lint-js-web/apache2/sites-available/001-www.quick-lint-js.com.conf [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+	RewriteEngine on
+	RewriteRule ^ https://quick-lint-js.com${DEV_TLD}%{REQUEST_URI} [END,NE,R=permanent]
+</VirtualHost>
+
+<VirtualHost *:443>
+	ServerName www.quick-lint-js.com${DEV_TLD}
+	RewriteEngine on
+	RewriteRule ^ https://quick-lint-js.com${DEV_TLD}%{REQUEST_URI} [END,NE,R=permanent]
+
+	SSLCertificateFile /etc/letsencrypt/live/quick-lint-js.com/fullchain.pem
```
Comment (strager) [inline infrastructure/quick-lint-js-web/apache2/sites-available/001-www.quick-lint-js.com.conf:12]: Isn't this cert file only for `quick-lint-js.com`, not for `www.quick-lint-js.com`?

Shouldn't we use a different cert for this domain:
```suggestion
 	SSLCertificateFile /etc/letsencrypt/live/www.quick-lint-js.com/fullchain.pem
```

Comment (strager) [reply infrastructure/quick-lint-js-web/apache2/sites-available/001-www.quick-lint-js.com.conf:12]: Nevermind. Certbot can create a cert for both domains in one file.

**Review by strager — COMMENTED** [review]
Note: Your patch will break `www.quicklintjs.com` (which I just added, so I don't blame you for breaking it).

**Comment by 70m4c** [comment]
Update: I have more updated and better tested Apache config as part of my Ansible project. Maybe I should close this draft PR and set up a new draft PR for that project.

**Comment by 70m4c** [comment]
I will replace this draft PR with a better PR soon™.

**Pull request closed without merging**